src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,138 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = new Node<>(null, null, null);
+        tail = head;
+        size = 0;
+    }
+
+    private void increaseSize(int value) {
+        size += value;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("The specified index is outside the list");
+        }
+    }
+
+    private Node<T> searchNode(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
```

Comment:
Don’t iterate your nodes from the head to tail or vice versa.

see common mistakes https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html